src/vcpkg/versions.cpp

```diff
@@ -50,93 +51,185 @@ namespace vcpkg::Versions
+            numbers.push_back(0);
+            cur = parse_skip_number(cur, &numbers.back());
+            if (!cur || (*cur != 0 && *cur != '.'))
+            {
+                return Strings::format(
+                    "Error: String `%s` must only contain dot-separated numeric values without leading zeroes.", str);
+            }
+            if (*cur == 0)
+            {
+                break;
+            }
+            // *cur == '.'
+            ++cur;
         }
 
-        SemanticVersion ret;
-        ret.original_string = str;
-        ret.version_string = str;
+        return RelaxedVersion{str, std::move(numbers)};
+    }
 
-        auto build_found = ret.version_string.find('+');
-        if (build_found != std::string::npos)
+    // 0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*
+    // equivalent:
+    static const char* skip_prerelease_identifier(const char* const s)
+    {
+        auto cur = s;
+        for (; Parse::ParserBase::is_ascii_digit(*cur); ++cur)
         {
-            ret.version_string.resize(build_found);
         }
-
-        auto prerelease_found = ret.version_string.find('-');
-        if (prerelease_found != std::string::npos)
+        const char ch = *cur;
+        if (Parse::ParserBase::is_lower_alpha(ch) || Parse::ParserBase::is_upper_alpha(ch) || ch == '-')
         {
-            ret.prerelease_string = ret.version_string.substr(prerelease_found + 1);
-            ret.identifiers = Strings::split(ret.prerelease_string, '.');
-            ret.version_string.resize(prerelease_found);
+            // matched alpha identifier
+            ++cur;
+            for (; Parse::ParserBase::is_alphanumdash(*cur); ++cur)
+            {
+            }
+            return cur;
         }
-
-        std::regex version_match("(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*)){2}");
-        if (!std::regex_match(ret.version_string, version_match))
+        if (*s == '0')
         {
-            return Strings::format("Error: String `%s` does not follow the required MAJOR.MINOR.PATCH format.",
-                                   ret.version_string);
+            // matches exactly zero
+            return s + 1;
         }
+        if (cur != s)
+        {
+            // matched numeric sequence (not starting with zero)
+            return cur;
+        }
+        return nullptr;
+    }
+
+    ExpectedS<SemanticVersion> SemanticVersion::from_string(const std::string& str)
+    {
+        // Suggested regex by semver.org
+        // std::regex semver_scheme_match("^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)"
+        //                                "(?:-((?:0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9][0-9]*|[0-9]"
+        //                                "*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
+        //                                "(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$");
+
+        SemanticVersion ret;
+        ret.original_string = str;
+        do
+        {
```

Comment:
Instead of this do-nothing-do in order to make break into poor-man's-goto, consider extracting a function so that you can use return for that purpose? e.g.:

```
bool from_string_impl(SemanticVersion& ret, const std::string& str) {
    ret.original_string = str;
    // the contents of the do loop go here
}

ExpectedS<SemanticVersion> SemanticVersion::from_string(const std::string& str) {
    SemanticVersion ret;
    if (from_string_impl(ret, str)) { return ret; }
    return Strings::format("Error: String `%s` is not a valid Semantic Version string, consult https://semver.org",
                                str);
}

Reply:
Imho it would be _much_ cleaner to just use `goto`, but I figured it would be less taxing on code review to avoid it.